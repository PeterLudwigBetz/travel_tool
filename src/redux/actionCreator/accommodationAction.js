import {
  CREATE_ACCOMMODATION_DATA,
  CREATE_ACCOMMODATION_DATA_SUCCESS,
  CREATE_ACCOMMODATION_DATA_FAILURE,
} from '../constants/actionTypes';

export const createAccommodation  = accommodationData => ({
  type: CREATE_ACCOMMODATION_DATA,
  accommodationData
});

export const createAccommodationSuccess = accommodationData => ({
  type: CREATE_ACCOMMODATION_DATA_SUCCESS,
  accommodationData,
});

export const createAccommodationFailure = error => ({
  type: CREATE_ACCOMMODATION_DATA_FAILURE,
  error,
});
