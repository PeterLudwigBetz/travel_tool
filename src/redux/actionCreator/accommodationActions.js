import {
  FETCH_ACCOMMODATION_CENTRES,
  FETCH_ACCOMMODATION_CENTRES_SUCCESS,
  FETCH_ACCOMMODATION_CENTRES_FAILURE,
  CREATE_ACCOMMODATION_DATA,
  CREATE_ACCOMMODATION_DATA_SUCCESS,
  CREATE_ACCOMMODATION_DATA_FAILURE,
} from '../constants/actionTypes';

export const fetchAccommodation = () => ({
  type: FETCH_ACCOMMODATION_CENTRES,
});

export const fetchAccommodationSuccess = ({guestHouses}) => ({
  type: FETCH_ACCOMMODATION_CENTRES_SUCCESS,
  guestHouses,
});

export const fetchAccommodationFailure = (error) => ({
  type: FETCH_ACCOMMODATION_CENTRES_FAILURE,
  error
});

export const createAccommodation  = accommodationData => ({
  type: CREATE_ACCOMMODATION_DATA,
  accommodationData
});

export const createAccommodationSuccess = accommodationData => ({
  type: CREATE_ACCOMMODATION_DATA_SUCCESS,
  accommodationData,
});

export const createAccommodationFailure = error => ({
  type: CREATE_ACCOMMODATION_DATA_FAILURE,
  error,
});
