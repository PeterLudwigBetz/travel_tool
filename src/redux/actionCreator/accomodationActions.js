import {
  INIT_FETCH_TIMELINE_DATA,
  FETCH_TIMELINE_DATA_SUCCESS
} from '../constants/actionTypes';

export const initFetchTimelineData = (guestHouseId, startDate, endDate) => ({
  type: INIT_FETCH_TIMELINE_DATA,
  guestHouseId,
  startDate,
  endDate
});

export const fetchTimelineDataSuccess = (guestHouse) => ({
  type: FETCH_TIMELINE_DATA_SUCCESS,
  guestHouse
});

export const fetchTimelineDataFailure = (errorMsg) => ({
  type: FETCH_TIMELINE_DATA_SUCCESS,
  guestHouse: {rooms: [] },
  error: errorMsg
});
