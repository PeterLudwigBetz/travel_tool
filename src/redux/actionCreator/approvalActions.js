import {
  FETCH_USER_APPROVALS,
  FETCH_USER_APPROVALS_SUCCESS,
  FETCH_USER_APPROVALS_FAILURE
} from '../constants/actionTypes';

export const fetchUserApprovals = (url) => ({ //eslint-disable-line
  type: FETCH_USER_APPROVALS,
  url
});


export const fetchUserApprovalsSuccess = (response) => ({
  type: FETCH_USER_APPROVALS_SUCCESS,
  approvals: response.approvals,
  message: response.message,
  meta:  response.meta
});

export const fetchUserApprovalsFailure = (error) => ({
  type: FETCH_USER_APPROVALS_FAILURE,
  error
});
