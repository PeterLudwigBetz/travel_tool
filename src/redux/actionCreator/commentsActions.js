import { CREATE_COMMENT, EDIT_COMMENT } from '../constants/actionTypes';

export const createComment = (requestId, comment) => ({
  type: CREATE_COMMENT,
  requestId,
  comment
});

export const createCommentSuccess = (comment) => ({
  type: 'CREATE_COMMENT_SUCCESS',
  comment
});

export const createCommentFailure = (error) => ({
  type: 'CREATE_COMMENT_FAILURE',
  error
});

export const editComment = (requestId, comment, id) => ({
  type: EDIT_COMMENT,
  requestId,
  comment,
  id
});

export const editCommentSuccess = (comment) => ({
  type: 'EDIT_COMMENT_SUCCESS',
  comment
});

export const editCommentFailure = (error) => ({
  type: 'EDIT_COMMENT_FAILURE',
  error
});
