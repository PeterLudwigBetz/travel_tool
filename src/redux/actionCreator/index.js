import {
  setCurrentUser,
  setCurrentUserSuccess,
  setCurrentUserFailure,
} from './actionCreators';
import {
  fetchUserApprovals,
  fetchUserApprovalsSuccess,
  fetchUserApprovalsFailure,
  updateRequestStatus,
  updateRequestStatusSuccess,
  updateRequestStatusFailure
} from './approvalActions';

export {
  setCurrentUser,
  setCurrentUserSuccess,
  setCurrentUserFailure,
  // Approvals actions
  fetchUserApprovals,
  fetchUserApprovalsSuccess,
  fetchUserApprovalsFailure,
  updateRequestStatus,
  updateRequestStatusSuccess,
  updateRequestStatusFailure
};
