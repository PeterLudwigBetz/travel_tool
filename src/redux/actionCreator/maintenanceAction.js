import { ADD_MAINTENANCE_RECORD } from '../constants/actionTypes';

const addMaintenenceRecord = (data, roomId) => ({
  type: ADD_MAINTENANCE_RECORD,
  data,
  roomId,
});

export default addMaintenenceRecord;
