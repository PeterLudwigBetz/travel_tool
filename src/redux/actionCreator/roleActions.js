import {
  FETCH_ROLE_USERS,
  FETCH_ROLE_USERS_SUCCESS,
  FETCH_ROLE_USERS_FAILURE,
} from '../constants/actionTypes';

export const fetchRoleUsers = roleId => ({
  type: FETCH_ROLE_USERS,
  roleId
});

export const fetchRoleUsersSuccess = ({roleName, users}) => ({
  type: FETCH_ROLE_USERS_SUCCESS,
  users,
  roleName
});

export const fetchRoleUsersFailure = error => ({
  type: FETCH_ROLE_USERS_FAILURE,
  error
});
