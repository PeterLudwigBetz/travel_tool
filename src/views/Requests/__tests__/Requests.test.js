import React from 'react';
import sinon from 'sinon';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { Requests, mapStateToProps } from '..';
import { fetchRoleUsers } from '../../../redux/actionCreator/roleActions';

const props = {
  requests: [
    {
      id: 'xDh20btGz',
      name: 'Amarachukwu Agbo',
      origin: 'Lagos',
      destination: 'Nairobi',
      manager: 'Samuel Kubai',
      gender: 'Female',
      department: 'TDD',
      role: 'Software Developer',
      status: 'Open',
      userId: 'pommyLHJmKrx76A8Slm',
      departureDate: '2018-12-09',
      arrivalDate: '2018-12-11',
    },
    {
      id: 'xDh20btGy',
      name: 'Amarachukwu Agbo',
      origin: 'Lagos',
      destination: 'Nairobi',
      manager: 'Samuel Kubai',
      gender: 'Female',
      department: 'TDD',
      role: 'Software Developer',
      status: 'Rejected',
      userId: 'pommyLHJmKrx76A8Slm',
      departureDate: '2018-12-09',
      arrivalDate: '2018-12-11',
    },
    {
      id: 'xDh20btGx',
      name: 'Amarachukwu Agbo',
      origin: 'Lagos',
      destination: 'Nairobi',
      manager: 'Samuel Kubai',
      gender: 'Female',
      department: 'TDD',
      role: 'Software Developer',
      status: 'Approved',
      userId: 'pommyLHJmKrx76A8Slm',
      departureDate: '2018-12-09',
      arrivalDate: '2018-12-11',
    },
  ],
  pagination: {
    currentPage: 1,
    pageCount: 4,
    dataCount: 10,
    onPageChange: sinon.spy(),
  },
  fetchUserRequests: sinon.spy(() => Promise.resolve()),
  fetchRoleUsers: sinon.spy(() => Promise.resolve()),
  fetchUserRequestsError: null,
  openRequestsCount: 1,
  pastRequestsCount: 1,
  isLoading: false,
  history: {
    push: jest.fn()
  },
  location: {
    search: '?page=1'
  },
  user: {
    UserInfo: {
      name: 'John Doe'
    }
  },
  createNewRequest: jest.fn(),
  loading: false,
  errors: [],
  shouldOpen: false,
  modalType: null,
  openModal: jest.fn(),
  closeModal: jest.fn(),
  page: 'Requests'
};

const initialState = {
  auth: {
    isAuthenticated: true,
    user: {
      UserInfo: {
        name: 'Tomato Jos',
        picture: 'http://picture.com/gif'
      }
    }
  },
  requestsReducer: {
    requests: [],
    request: {},
    loading: false,
    errors: []
  },
  modalReducer: {
    shouldOpen: false,
    modalType: null
  },
  getCurrentUserRole: 'tomato'
};
const mockStore = configureStore();
const store = mockStore(initialState);


describe('<Requests>', () => {
  it('should render the Requests without crashing', () => {
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Requests {...props} />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.length).toBe(1);
    wrapper.unmount();
  });

  it('calls the componentDidMount method', () => {
    const spy = sinon.spy(Requests.prototype, 'componentDidMount');
    const { fetchUserRequests, fetchRoleUsers } = props;
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Requests {...props} />
        </MemoryRouter>
      </Provider>
    );
    expect(spy.called).toEqual(true);
    expect(fetchUserRequests.called).toEqual(true);
    expect(fetchUserRequests.calledWith('?page=1')).toEqual(true);
    expect(fetchRoleUsers.called).toEqual(true);
    expect(fetchRoleUsers.calledWith(53019)).toEqual(true);
    wrapper.unmount();
  });

  it(`calls the fetchRequests method
    when the next pagination button is clicked`, () => {
    const { fetchUserRequests } = props;
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Requests {...props} />
        </MemoryRouter>
      </Provider>
    );
    const spy = sinon.spy(wrapper.find(Requests).instance(), 'fetchRequests');
    wrapper.find('#next-button').simulate('click');
    expect(spy.calledOnce).toEqual(true);
    expect(fetchUserRequests.called).toEqual(true);
    wrapper.unmount();
  });

  it(`calls the fetchRequests method with the correct query
    when the next pagination button is clicked`, () => {
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Requests {...{
            ...props,
            location:{
              search: ''
            }
          }} />
        </MemoryRouter>
      </Provider>
    );
    const spy = sinon.spy(wrapper.find(Requests).instance(), 'fetchRequests');
    wrapper.find('#next-button').simulate('click');
    expect(spy.called).toEqual(true);
    expect(spy.calledWith('?page=2')).toEqual(true);
    wrapper.unmount();
  });

  it(`calls the fetchRequests method
    with the selected limit when a limit is selected`, () => {
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Requests {...props} />
        </MemoryRouter>
      </Provider>
    );
    const spy = sinon.spy(wrapper.find(Requests).instance(), 'fetchRequests');
    wrapper.find('.dropdown__list__item').first().simulate('click');
    expect(spy.calledOnce).toEqual(true);
    expect(spy.calledWith('?page=1&limit=10')).toEqual(true);
    wrapper.unmount();
  });

  it(`calls the fetchRequests method with the last possible page
  when a limit and a page above the available pages is selected`, () => {
    const { pagination } = props;
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Requests {...{
            ...props,
            pagination: {...pagination, currentPage: 4},
            location: { search: '?page=4&status=open'}
          }} />
        </MemoryRouter>
      </Provider>
    );
    const spy = sinon.spy(wrapper.find(Requests).instance(), 'fetchRequests');
    wrapper.find('.dropdown__list__item').last().simulate('click');
    expect(spy.calledOnce).toEqual(true);
    expect(spy.calledWith('?page=1&status=open&limit=30')).toEqual(true);
    wrapper.unmount();
  });

  it('should render the child components', () => {
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Requests {...props} />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find('.rp-requests__header').length).toBe(1);// RequestsPanelHeader
    expect(wrapper.find('Table').length).toBe(1);
    expect(wrapper.find('RequestPanelHeader').length).toBe(1);
    expect(wrapper.find('WithLoading').length).toBe(1); //WithLoading HOC containing Requests
    // Since the element is always on the DOM, the page length will always be one
    expect(wrapper.find('Pagination').length).toBe(1);
    wrapper.unmount();
  });

  it('should call renderRequestPanelHeader method', () =>{
    const wrapper = shallow(<Requests {...props} />);
    const instance = wrapper.instance();
    const spyon = jest.spyOn(instance, 'renderRequestPanelHeader');
    expect(spyon).toHaveBeenCalledTimes(0);
  });

  it('renders as expected', () => {
    const wrapper = shallow(<Requests {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it.skip('should set `shouldOpen` prop to `true` when new request button is clicked', () => {
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Requests {...props} />
        </MemoryRouter>
      </Provider>
    );
    wrapper
      .find('.btn-new-request')
      .simulate('click');
    expect(wrapper.find('Requests').props().shouldOpen).toEqual(true);
  });

  it('should set `visibility` prop to `visible` when new request button is clicked', () => {
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Requests {...{...props, shouldOpen: true, modalType: 'new model'}} />
        </MemoryRouter>
      </Provider>
    );
    wrapper
      .find('.btn-new-request')
      .simulate('click');
    expect(wrapper.find('Modal').at(0).props().visibility).toEqual('visible');
  });

  it('maps state to props and return the expected object', () => {
    const requests = {
      requests: [],
      errors: [],
      loading: false,
    };
    const modal = {
      modal: {
        shouldOpen: false,
        modalType: null,
      }
    };
    const props = mapStateToProps({requests, modal});
    expect(props).toEqual({...requests, ...modal.modal});
  });
});
