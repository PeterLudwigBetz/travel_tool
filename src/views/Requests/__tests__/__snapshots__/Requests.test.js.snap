// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Requests> renders as expected 1`] = `
<Fragment>
  <Modal
    closeModal={[Function]}
    modalBar={<div />}
    modalContentId=""
    modalId=""
    params=""
    title="New Travel Request"
    visibility="invisible"
    width="1000px"
  >
    <NewRequestForm
      closeModal={[Function]}
      creatingRequest={false}
      errors={Array []}
      handleCreateRequest={[Function]}
      loading={false}
      managers={Array []}
      updateUserProfile={[Function]}
      user={
        Object {
          "UserInfo": Object {
            "name": "John Doe",
          },
        }
      }
    />
  </Modal>
  <div
    className="rp-requests__header"
  >
    <RequestPanelHeader
      activeStatus="all"
      fetchRequests={[Function]}
      getRequestsWithLimit={[Function]}
      modalType={null}
      openModal={[Function]}
      openRequestsCount={1}
      pastRequestsCount={1}
      requestsLength={3}
      shouldOpen={false}
      url="?page=1"
    />
  </div>
  <div
    className="rp-table"
  >
    <WithLoading
      closeModal={[Function]}
      fetchRequestsError={null}
      isLoading={false}
      modalType={null}
      openModal={[Function]}
      page="Requests"
      requests={
        Array [
          Object {
            "department": "TDD",
            "gender": "Female",
            "id": "xDh20btGz",
            "manager": "Ezrqn Kiptanui",
            "name": "Amarachukwo Agbo",
            "role": "Learning Facilitator",
            "tripType": "oneWay",
            "trips": Array [
              Object {
                "departureDate": "2018-09-20",
                "destination": "Angola",
                "origin": "Lagos",
              },
            ],
          },
          Object {
            "department": "TDD",
            "gender": "Female",
            "id": "xDh20btGy",
            "manager": "Ezrqn Kiptanui",
            "name": "Amarachukwo Agbo",
            "role": "Learning Facilitator",
            "tripType": "oneWay",
            "trips": Array [
              Object {
                "departureDate": "2018-09-20",
                "destination": "Angola",
                "origin": "Lagos",
              },
            ],
          },
          Object {
            "department": "TDD",
            "gender": "Female",
            "id": "xDh20btGx",
            "manager": "Ezrqn Kiptanui",
            "name": "Amarachukwo Agbo",
            "role": "Learning Facilitator",
            "tripType": "oneWay",
            "trips": Array [
              Object {
                "departureDate": "2018-09-20",
                "destination": "Angola",
                "origin": "Lagos",
              },
            ],
          },
        ]
      }
      shouldOpen={false}
      type="requests"
    />
  </div>
  <div
    className="rp-pagination"
  >
    <Pagination
      currentPage={1}
      onPageChange={[Function]}
      pageCount={4}
    />
  </div>
</Fragment>
`;
